### PR TITLE
Feature/audit package

### PR DESCRIPTION
**Pull request by micahklaz**
Initial commits while working on tests

### CODE REVIEW

File: packages/ap-audit/docs/api.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+Middleware
+**********
+
+.. autoclass:: alliance_platform.audit.middleware.AuditMiddleware
+
+
+Templatetags
+************
+
+.. autofunction:: alliance_platform.audit.templatetags.alliance_platform.audit.render_audit_list
```
Comment (davecoates) [inline packages/ap-audit/docs/api.rst:51]: using autofunction for templatetags sucks :( so now templatetags docs should be written like:

```
.. templatetag:: render_audit_list
                                                                                                                     
``render_audit_list``
---------------------

description here... duplication from the tag itself unfortunately (or you can have a dev friendly header doc
in the templatetag and user friendly docs here)
                                  
================= =============================================================
Argument          Description
================= =============================================================
``arg1``          blah blah
``**kwargs``      Any other props to pass through to the audit component
================= =============================================================                                                                                             
```

(you can see examples in templatetags.rst in frontend package)

You can then link to templatetags with just the tagname like 

```
:ttag:`render_audit_list`
```

File: packages/ap-audit/docs/installation.rst [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+If you want to write your own audit log component, you can use the ``useAuditEndpoint`` hook from
+``@alliancesoftware/ui-audit`` to help interface between your component and the audit log endpoint.
+This will work out of the box with the props passed from the ``render_audit_list`` templatetab.
+
+Audit Template Tags
+~~~~~~~~~~~~~~~~~~~
+
+To use the audit list templatetags, you need to install `alliance_platform.frontend <http://127.0.0.1:56676/>`__ and add it to ``INSTALLED_APPS``.
+
+You will also need to ensure that the ``AUDIT_LOG_COMPONENT_PATH`` setting points to a React component that will render
```
Comment (davecoates) [inline packages/ap-audit/docs/installation.rst:41]: link to setting, e.g. 
```
:data:`~alliance_platform.audit.settings.AlliancePlatformAuditSettingsType.AUDIT_LOG_COMPONENT_PATH` 
```

File: packages/ap-audit/docs/installation.rst [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+templatetag looks in this package to import the audit log component to display.
+
+If you want to write your own audit log component, you can use the ``useAuditEndpoint`` hook from
+``@alliancesoftware/ui-audit`` to help interface between your component and the audit log endpoint.
+This will work out of the box with the props passed from the ``render_audit_list`` templatetab.
+
+Audit Template Tags
+~~~~~~~~~~~~~~~~~~~
+
+To use the audit list templatetags, you need to install `alliance_platform.frontend <http://127.0.0.1:56676/>`__ and add it to ``INSTALLED_APPS``.
```
Comment (davecoates) [inline packages/ap-audit/docs/installation.rst:39]: can't use localhost, should be 
```
:doc:`alliance_platform_frontend <alliance-platform-frontend:installation>`
```

File: packages/ap-audit/docs/installation.rst [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+* Add :class:`alliance_platform.audit.middleware.AuditMiddleware` to :setting:`MIDDLEWARE`
+
+* In ``urls.py``, add paths for :class:`alliance_platform.audit.api.AuditLogView` and :class:`alliance_platform.audit.api.AuditUserChoicesView`.
+
+Frontend
+~~~~~~~~
+
+If you want to use the pre-made audit log component (built off of ``@alliancesoftware/ui``), you will need
+to add ``@alliancesoftware/ui-audit`` to your ``package.json``. By default, the ``render_audit_list``
```
Comment (davecoates) [inline packages/ap-audit/docs/installation.rst:29]:
```
:ttag:`render_audit_list`
```

(but only will work after doc changes suggested for api.rst)

File: packages/ap-audit/docs/installation.rst [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+templatetag looks in this package to import the audit log component to display.
+
+If you want to write your own audit log component, you can use the ``useAuditEndpoint`` hook from
+``@alliancesoftware/ui-audit`` to help interface between your component and the audit log endpoint.
+This will work out of the box with the props passed from the ``render_audit_list`` templatetab.
```
Comment (davecoates) [inline packages/ap-audit/docs/installation.rst:34]: templatetab => templatetag

File: packages/ap-audit/alliance_platform/audit/settings.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    .. code-block:: python
+
+        ALLIANCE_PLATFORM = {
+            "AUDIT": {
+                # settings go here
+            }
+        }
+    """
+
+    #: The name of the action used to evaluate permissions for auditing a model. Defaults to 'audit'.
```
Comment (davecoates) [inline packages/ap-audit/alliance_platform/audit/settings.py:23]: Can you add to the end

```
The action name is passed through :func:`~alliance_platform.core.auth.resolve_perm_name` to generate the permission.
```

File: packages/ap-audit/alliance_platform/audit/settings.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    LIST_PERM_ACTION: str | None
+    #: The name of the permission used to evaluate whether auditing of any type can be performed. If it resolves to ``False``,
+    #: this will override individual audit permissions determined by ``LIST_PERM_ACTION``. Defaults to
```
Comment (davecoates) [inline packages/ap-audit/alliance_platform/audit/settings.py:26]: I don't understand this comment about resolving to False. Type doesn't accept `False`, and it doesn't look like that would do anything anyway?

File: packages/ap-audit/docs/usage.rst
Comment (davecoates) [inline packages/ap-audit/docs/usage.rst:1]: could you go through this file and fix up the doc links, a lot of the references don't link

Comment (davecoates) [reply packages/ap-audit/docs/usage.rst:1]: actually just load the docs up locally, and go through each page - most pages have references that don't link

File: packages/ap-audit/alliance_platform/audit/settings.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    #: ``"alliance_platform.audit.can_audit"``
+    GLOBAL_AUDIT_PERMISSION_NAME: str | None
+    #: The SQL function used to generate the display name for users in audit logs. Defaults to showing first name and last name
+    USERNAME_FORMAT: Func | None
+    #: The format string to use when displaying timestamps for audit logs. Defaults to ``"%Y-%m-%d %H:%M:%S"``
+    DATETIME_FORMAT: str | None
```
Comment (davecoates) [inline packages/ap-audit/alliance_platform/audit/settings.py:32]: hmm why do we need to format it here? don't we just return the date in the API, so couldn't it be a standard date string? (maybe i'm wrong about that, haven't checked the UI code)

File: packages/ap-audit/alliance_platform/audit/settings.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    """
+
+    #: The name of the action used to evaluate permissions for auditing a model. Defaults to 'audit'.
+    LIST_PERM_ACTION: str | None
+    #: The name of the permission used to evaluate whether auditing of any type can be performed. If it resolves to ``False``,
+    #: this will override individual audit permissions determined by ``LIST_PERM_ACTION``. Defaults to
+    #: ``"alliance_platform.audit.can_audit"``
+    GLOBAL_AUDIT_PERMISSION_NAME: str | None
+    #: The SQL function used to generate the display name for users in audit logs. Defaults to showing first name and last name
```
Comment (davecoates) [inline packages/ap-audit/alliance_platform/audit/settings.py:29]: could you include the actual code so it's more obvious what you can actually set here: `Concat("first_name", Value(" "), "last_name")`. Add a warning that if your user model doesn't include first_name + last_name fields you have to set this setting

File: packages/ap-audit/alliance_platform/audit/settings.py (hunk 8 above)
Comment (micahklaz) [reply packages/ap-audit/alliance_platform/audit/settings.py:32]: I'm not entirely sure how useful this is. It carried over from `common_audit`. The date is converted to a string for serialisation specifically.

**Pull request merged**